core/src/main/java/org/apache/brooklyn/core/typereg/BasicBrooklynTypeRegistry.java

```diff
@@ -368,13 +399,77 @@ public void addToLocalUnpersistedTypeRegistry(RegisteredType type, boolean canFo
+                        if (canForce) {
+                            log.debug("Addition of "+type+" to replace "+existingT+" allowed because force is on");
+                            continue;
+                        }
+                        if (BrooklynVersionSyntax.isSnapshot(type.getVersion())) {
+                            if (existingT.getContainingBundle()!=null) {
+                                if (BrooklynVersionSyntax.isSnapshot(VersionedName.fromString(existingT.getContainingBundle()).getVersionString())) {
+                                    log.debug("Addition of "+type+" to replace "+existingT+" allowed because both are snapshot");
+                                    continue;
+                                } else {
+                                    reasonForDetailedCheck = "the containing bundle "+existingT.getContainingBundle()+" is not a SNAPSHOT and addition is not forced";
+                                }
+                            } else {
+                                // can this occur?
+                                reasonForDetailedCheck = "the containing bundle of the type is unknown (cannot confirm it is snapshot)";
+                            }
+                        } else {
+                            reasonForDetailedCheck = "the type is not a SNAPSHOT and addition is not forced";
+                        }
+                    } else if (oldIsWrapperBundle) {
+                        reasonForDetailedCheck = type.getId()+" is in a named bundle replacing an item from an anonymous bundle-wrapped BOM, so definitions must be the same (or else give it a different version)";
+                    } else if (newIsWrapperBundle) {
+                        reasonForDetailedCheck = type.getId()+" is in an anonymous bundle-wrapped BOM replacing an item from a named bundle, so definitions must be the same (or else give it a different version)";
+                    } else {
+                        reasonForDetailedCheck = type.getId()+" is defined in different bundle";
+                    }
+                    assertSameEnoughToAllowReplacing(existingT, type, reasonForDetailedCheck);
+                }
+            
+                log.debug("Inserting "+type+" into "+this+
+                    (oldContainingBundlesToRemove.isEmpty() ? "" : " (removing entry from "+oldContainingBundlesToRemove+")"));
+                for (String oldContainingBundle: oldContainingBundlesToRemove) {
+                    knownMatchingTypesByBundles.remove(oldContainingBundle);
+                }
+                knownMatchingTypesByBundles.put(type.getContainingBundle(), type);
+            });
+    }
+
+    private boolean isWrapperBundle(String bundleNameVersion) { 
+        if (bundleNameVersion==null) return true;
+        Maybe<OsgiManager> osgi = ((ManagementContextInternal)mgmt).getOsgiManager();
+        // if not osgi, everything is treated as a wrapper bundle
+        if (osgi.isAbsent()) return true;
```

Comment:
[bigger topic than this PR, but raising it here while I'm thinking about it].

Now that osgi is the "only" way to run brooklyn (except for tests, and for developers using a quick launch in their IDE?), I wonder how we want to handle `osgi.isAbsent()` long term. For example, instead of handling it in our code in lots of places, perhaps tests should inject a pseudo-osgi-manager that just tracks bundle install/delete/find, and does a simple peek inside it for a catalog.bom at the root. Anything more complicated would fail-fast.

Let's discuss that separately, but it does feel like a way to simplify the production code (making our test framework code slightly more complicated instead).

Reply:
that would be nice.  yes, pseudo-osgi `OsgiManager` might be cleaner than checks throughout our code, depending how much smarts are needed there.  alternatively it could spin up felix on demand.

other thing to consider is _only_ osgi, though that make make running individual IDE tests too cumbersome to set up and/or slow to run.